drivers/clock_control/nrf_clock_calibration.c

```diff
@@ -51,124 +47,115 @@ static void cal_hf_on_callback(struct device *dev,
+				&cal_lf_on_data);
 }
 
-bool z_nrf_clock_calibration_stop(struct device *dev)
+static void lf_release(void)
 {
-	int key;
-	bool ret = true;
-
-	key = irq_lock();
+	clock_control_off(clk_dev, CLOCK_CONTROL_NRF_SUBSYS_LF);
+}
 
-	nrf_clock_task_trigger(NRF_CLOCK, NRF_CLOCK_TASK_CTSTOP);
-	nrf_clock_event_clear(NRF_CLOCK, NRF_CLOCK_EVENT_CTTO);
+static void cal_lf_on_callback(struct device *dev,
+				clock_control_subsys_t subsys, void *user_data)
+{
+	hf_request();
+}
 
-	/* If calibration is active then pend until completed.
-	 * Currently (and most likely in the future), LFCLK is never stopped so
-	 * it is not an issue.
-	 */
-	if (cal_state == CAL_ACTIVE) {
-		cal_state = CAL_ACTIVE_OFF;
-		ret = false;
-	} else {
-		cal_state = CAL_OFF;
+/* Start actual HW calibration assuming that HFCLK XTAL is on. */
+static void start_hw_cal(void)
+{
+	/* Workaround for Errata 192 */
+	if (IS_ENABLED(CONFIG_SOC_SERIES_NRF52X)) {
+		*(volatile uint32_t *)0x40000C34 = 0x00000002;
 	}
 
-	irq_unlock(key);
-	LOG_DBG("Stop requested %s.", (cal_state == CAL_ACTIVE_OFF) ?
-			"during ongoing calibration" : "");
+	nrf_clock_task_trigger(NRF_CLOCK, NRF_CLOCK_TASK_CAL);
+	calib_skip_cnt = CONFIG_CLOCK_CONTROL_NRF_CALIBRATION_MAX_SKIP;
+}
 
-	return ret;
+/* Start cycle by starting backoff timer and releasing HFCLK XTAL. */
+static void start_cycle(void)
+{
+	k_timer_start(&backoff_timer,
+		      K_MSEC(CONFIG_CLOCK_CONTROL_NRF_CALIBRATION_PERIOD), 0);
+	hf_release();
+	lf_release();
+	cal_process_in_progress = 0;
 }
 
-void z_nrf_clock_calibration_init(struct device *dev)
+static void start_cal_process(void)
 {
-	/* Anomaly 36: After watchdog timeout reset, CPU lockup reset, soft
-	 * reset, or pin reset EVENTS_DONE and EVENTS_CTTO are not reset.
+	if (atomic_cas(&cal_process_in_progress, 0, 1) == false) {
+		return;
+	}
+
+	/* LF clock is probably running but it is requested to ensure that
+	 * it is not released while calibration process in ongoing. If system
+	 * releases the clock during calibration process it will be released
+	 * at the end of calibration process and stopped in consequence.
 	 */
-	nrf_clock_event_clear(NRF_CLOCK, NRF_CLOCK_EVENT_DONE);
-	nrf_clock_event_clear(NRF_CLOCK, NRF_CLOCK_EVENT_CTTO);
+	lf_request();
+}
 
-	nrf_clock_int_enable(NRF_CLOCK, NRF_CLOCK_INT_DONE_MASK |
-					NRF_CLOCK_INT_CTTO_MASK |
-					NRF_CLOCK_INT_CTSTOPPED_MASK);
-	nrf_clock_cal_timer_timeout_set(NRF_CLOCK,
-			CONFIG_CLOCK_CONTROL_NRF_CALIBRATION_PERIOD);
+static void timeout_handler(struct k_timer *timer)
+{
+	start_cal_process();
+}
 
-	if (CONFIG_CLOCK_CONTROL_NRF_CALIBRATION_MAX_SKIP != 0) {
-		temp_sensor = device_get_binding(TEMP_SENSOR_NAME);
+/* Called when HFCLK XTAL is on. Schedules temperature measurement or triggers
+ * calibration.
+ */
+static void cal_hf_on_callback(struct device *dev,
+				clock_control_subsys_t subsys, void *user_data)
+{
+	if ((CONFIG_CLOCK_CONTROL_NRF_CALIBRATION_MAX_SKIP == 0) ||
```

Comment:
I think the first condition should instead be `temp_sensor != NULL`.  Initialization of `temp_sensor` is conditional on `CONFIG_CLOCK_CONTROL_NRF_CALIBRATION_MAX_SKIP != 0`, so in the normal case the conditions are equivalent.  But if something goes wrong in in the temperature sensor driver such that `device_get_binding()` returns NULL then `get_temperature()` will start to use it which at worst will cause a fault, and at best return an error that will prevent calibration from proceeding.  In either case running the calibration without checking temperature is a better fallback.

Reply:
changed `temp_sensor == NULL` (I think that what you meant)